docs/proposals/20210524-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,582 @@
+
+
+## Motivation
+
+- Motivation behind having cluster-api provision and manage the external etcd cluster
+
+  - Cluster API supports the use of an external etcd cluster, by allowing users to provide their external etcd cluster's endpoints.
+  So it would be good to add support for provisioning the etcd cluster too, so the user can have a single workflow for full management of a cluster using the external etcd topology. The user can offload the responsibility of managing an external etcd cluster to CAPI.
+    
+  - External etcd topology decouples the control plane and etcd member. So if a control plane-only node fails, or if there's a memory leak in a component like kube-apiserver, it won't directly impact an etcd member.
+    
+  - Etcd is resource intensive, so it is safer to have dedicated nodes for etcd, it could use more disk space, higher bandwidth. 
+    Having a separate etcd cluster for these reasons could ensure a more resilient HA setup.
+    
+- Motivation behind using etcdadm as a pluggable etcd provider
+  
+  - Leveraging existing projects such as etcdadm is one way of bringing up an etcd cluster. One of the goals mentioned in the [etcdadm project roadmap](https://github.com/kubernetes-sigs/etcdadm/blob/master/ROADMAP.md) is to integrate it with CAPI to support the "external etcd" use case.
+  
+  - Etcd providers can be made pluggable and we can start by adding one that uses etcdadm.
+
+### Goals
+
+- Introduce pluggable etcd providers in Cluster API, starting with an etcdadm based provider. This etcdadm provider will create and manage an etcd cluster.
+- User should be able to create a Kubernetes cluster that uses external etcd topology in a single step.
+- User should only have to create the required CAPI resources, along with the required etcd provider specific resources. That should trigger creation of an etcd cluster, followed by creation of the target workload cluster which uses this etcd cluster. Meaning, the CAPI controllers will deploy the etcd cluster first and then use that as input to the workload cluster, without requiring any user intervention.
+- There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
+- Define a contract for pluggable etcd providers and define steps that control plane providers should take to use a managed external etcd cluster.
+- Support the following etcd cluster management actions: scale up and scale down, etcd member replacement and etcd version upgrades.
+- The etcd providers will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.
+
+### Non-Goals
+- The first iteration will use IP addresses/hostnames as etcd cluster endpoints. It can not configure static endpoint(s) till the Load Balancer provider is available.
```

Comment:
Uhmm, would this trigger a control plane rollout every time the list of address/host names changes?

Reply:
Yes, it would trigger a control plane rollout in this first iteration since we decided to configure static endpoints through the load balancer provider when it is implemented. Not sure if there is a different way of updating the kube-apiserver, with the etcd endpoints, can't think of one. And for static endpoints can't think of a different approach that won't need changes in the infrastructure providers (apart from the load balancer provider). Ideally the etcd endpoints should only change when etcd machines are replaced, in case of upgrades or if a member is unhealthy and needs to be replaced.

Reply:
What about asking the users to provide a stable address/a dns for the etcd cluster?

Reply:
yes I have included that in the "Future work" section. In that I've also included how using multiple dns records as static endpoints, one per etcd member is beneficial because through that we can leverage [etcd client's balancer](https://etcd.io/docs/v3.4/learning/design-client/#clientv3-grpc10-balancer-overview).
We discussed in the last CAPI meeting how the contract shouldn't change once we start using static endpoints. Right now the contract section in this KEP specifies that the etcd cluster's endpoint is a list of IP addresses for the first iteration. Later on when we use DNS records as static endpoints, if we configure a dns record per member (to leverage etcd's balancer) we will still have a list of endpoints, so nothing has to change in the contract.